File: spigot-impl-hidden-details/src/main/java/us/minevict/mvutil/common/utils/SetupPlatformless.java
package us.minevict.mvutil.common.utils;

import co.aikar.commands.BukkitCommandIssuer;
import co.aikar.commands.CommandIssuer;
import java.util.concurrent.CompletableFuture;
import java.util.function.BiConsumer;
import java.util.function.Function;
import java.util.function.Supplier;
import net.md_5.bungee.api.chat.BaseComponent;
import org.bukkit.plugin.Plugin;
import org.bukkit.scheduler.BukkitRunnable;

/**
 * @since 3.6.0
 */
public class SetupPlatformless {
  private SetupPlatformless() {
  }

  /**
   * Set up the {@link Platformless} functionality.
   * <p>
   * This <b>must</b> be called (at least) once.
   * </p>
   *
   * @param plugin The plugin executing this.
   */
  public static void setup(Plugin plugin) {
    Platformless.messageSender = PlatformlessMessageSenderSpigot.INSTANCE;
    Platformless.asyncRunner = new PlatformlessAsyncRunnerSpigot(plugin);
  }

  public static class PlatformlessMessageSenderSpigot implements
      BiConsumer<CommandIssuer, BaseComponent[]> {
    public static final PlatformlessMessageSenderSpigot INSTANCE = new PlatformlessMessageSenderSpigot();

    private PlatformlessMessageSenderSpigot() {
    }

    @Override
    public void accept(CommandIssuer commandIssuer, BaseComponent[] baseComponents) {
      if (baseComponents == null || !(commandIssuer instanceof BukkitCommandIssuer)) {
        return;
      }

      ((BukkitCommandIssuer) commandIssuer).getIssuer().sendMessage(baseComponents);
    }
  }

  public static class PlatformlessAsyncRunnerSpigot implements
      Function<Supplier, CompletableFuture> {
    private final Plugin plugin;

    public PlatformlessAsyncRunnerSpigot(Plugin plugin) {
      this.plugin = plugin;
    }

    @SuppressWarnings({"rawtypes", "unchecked"})
    @Override
    public CompletableFuture apply(Supplier supplier) {
      CompletableFuture future = new CompletableFuture();
      new BukkitRunnable() {
        @Override
        public void run() {
          try {
            future.complete(supplier.get());
          } catch (Throwable throwable) {
            future.completeExceptionally(throwable);
          }
        }
      }.runTaskAsynchronously(plugin);
      return future;
    }
  }
}

File: spigot/src/main/kotlin/us/minevict/mvutil/spigot/internal/SetupPlatformless.kt
/**
 * MV-Util
 * Copyright (C) 2020 Mariell Hoversholm, Nahuel Dolores
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package us.minevict.mvutil.spigot.internal

import co.aikar.commands.BukkitCommandIssuer
import org.bukkit.plugin.Plugin
import org.bukkit.scheduler.BukkitRunnable
import us.minevict.mvutil.common.utils.Platformless
import us.minevict.mvutil.spigot.MinevictusUtilsSpigot
import us.minevict.mvutil.spigot.ext.copyResource
import java.util.concurrent.CompletableFuture

/**
 * @since 5.0.0
 */
internal object SetupPlatformless {
    fun setup(plugin: MinevictusUtilsSpigot) {
        Platformless.messageSender = { issuer, components ->
            if (issuer is BukkitCommandIssuer) {
                issuer.player.sendMessage(*components)
            }
        }
        Platformless.asyncRunner = {
            val future = CompletableFuture<Any?>()
            object : BukkitRunnable() {
                override fun run() {
                    try {
                        future.complete(it())
                    } catch (ex: Throwable) {
                        future.completeExceptionally(ex)
                    }
                }
            }.runTaskAsynchronously(plugin)
            future
        }
        Platformless.exportJarResource = b@{ plug, name, file ->
            if (plug is Plugin) plug.copyResource(name, file)
            else false
        }
    }
}